### ISSUE TITLE
copy_to guesses column types better with RMySQL than RMariaDB  

### ISSUE THREAD

**Issue by ericemc3** [r-dbi/RMariaDB#163]
With RMySQL, a "copy_to" exports a table by following the types and lengths of its columns quite precisely, for instance adjusting the type and computing the proper length for VARCHAR cols. It is really handy! 

With RMariaDB, the character columns are converted approximately (they are actually converted to blobs).

Could RMariaDB catch up with RMySQL in this respect?

Because it is quite painful to have to explicitly specify field.types (dbWriteTable) for all columns. Of course i could go on with RMySQL, but it is apparently no longer supported.

```r
library(tidyverse) 
library(RMySQL) 
library(RMariaDB) 

dt <- tribble(
  ~lib, ~id, ~x,
  "Bage-le-Chatel",     "01", 3,
  "Asnieres-sur-Saone", "02", 1,
  "Asnieres",           "03", 1,
)

# RMySQL
db_cred = list( drv = MySQL(), host="xxx", dbname="xxx", user="xxx", password="xxx" )
mycon <- exec( "dbConnect", !!!db_cred ) 

copy_to( mycon, dt, name = 'test', overwrite = T, temporary = FALSE ) 

dbExecute( mycon, "ALTER TABLE test ADD PRIMARY KEY(id)" )
rs <- dbSendQuery( mycon, "SHOW COLUMNS FROM test" )
dbFetch(rs)

##  Field        Type Null Key Default Extra
##1   lib varchar(18)  YES        <NA>      
##2    id  varchar(2)   NO PRI              
##3     x      double  YES        <NA> 

# RMariaDB
db_cred = list( drv = MariaDB(), host="xxx", dbname="xxx", user="xxx", password="xxx" )
mycon <- exec( "dbConnect", !!!db_cred ) 

copy_to( mycon, dt, name = 'test', overwrite = T, temporary = FALSE ) 

dbExecute( mycon, "ALTER TABLE test ADD PRIMARY KEY(id)" )
# Error : BLOB/TEXT column 'id' used in key specification without a key length [1170]

rs <- dbSendQuery( mycon, "SHOW COLUMNS FROM test" )
dbFetch(rs)

##  Field   Type Null Key Default Extra
##1   lib   text  YES        <NA>      
##2    id   text  YES        <NA>      
##3     x double  YES        <NA>    
```

**Comment by krlmlr** [comment]
Thanks. Is this a dbplyr problem?

**Comment by krlmlr** [comment]
Does #159 fix this problem?

**Comment by ericemc3** [comment]
The problem is that RMariaDB behaves less efficiently than RMySQL. Of course i know that column of type text cannot use indexes, and i don"t want RMariaDB to output blob columns instead of varchar(2) columns.

My demand is: is it possible to improve RMariaDB so that it behaves better, and at least at the same level of efficiency than RMySQL, since it is promoted as a replacement for RMySQL ?

I really think it is dbplyr interest to benefit from a better MySQL support within RMariaDB

**Comment by krlmlr** [comment]
I hear you, I agree we shouldn't use `text` columns. Does #159 fix this problem?

**Comment by ericemc3** [comment]
I don't think so.
I would like `copy_to` here:
```r
# RMariaDB
db_cred = list( drv = MariaDB(), host="xxx", dbname="xxx", user="xxx", password="xxx" )
mycon <- exec( "dbConnect", !!!db_cred ) 

copy_to( mycon, dt, name = 'test', overwrite = T, temporary = FALSE ) 
```
to create a mysql table identical to the source dataframe (varchar(18), varchar(2) column...), the same way RMySQL does.

**Comment by krlmlr** [comment]
It might, though. Have you tried `remotes::install_github("r-dbi/RMariaDB#159")` and rerunning your code (after restarting R)?

**Comment by ericemc3** [comment]
Sorry, i hadn't understood your suggestion (to install a new dev version).

So i just did, and it works as expected, great, thank you!

**Comment by ericemc3** [comment]
By the way, i read in #159  (table.R):

```r
get_char_type <- function(x) {
  width <- max(max(nchar(enc2utf8(x)), na.rm = TRUE), 1)
  if (width > 255) {
    "TEXT"
  } else {
    paste0("VARCHAR(", width, ")")
  }
}
```
Not sure if it makes sense, but actually a VARCHAR column has a 65,535 max-length with MySQL (5+), so the > 255 for TEXT type seems to me still a little tough.

**Comment by ericemc3** [comment]
Hello is the #159 version soon becoming the new official release ?
Thanks

**Comment by krlmlr** [comment]
Thanks for your input. I merged #159, releasing to CRAN now.

**Comment by ericemc3** [comment]
Thank you very much !
